dbms/src/Flash/Mpp/MPPTunnelSetWriter.cpp

```diff
@@ -30,52 +30,159 @@ void checkPacketSize(size_t size)
+        switch (compress_method)                                                            \
+        {                                                                                   \
+        case CompressionMethod::NONE:                                                       \
+        {                                                                                   \
+            UPDATE_EXCHANGE_MATRIC_NONE_COMPRESS(type, is_local, actual_size);              \
+            break;                                                                          \
+        }                                                                                   \
+        case CompressionMethod::LZ4:                                                        \
+        {                                                                                   \
+            UPDATE_EXCHANGE_MATRIC_LZ4_COMPRESS(type, actual_size);                         \
+            break;                                                                          \
+        }                                                                                   \
+        case CompressionMethod::ZSTD:                                                       \
+        {                                                                                   \
+            UPDATE_EXCHANGE_MATRIC_ZSTD_COMPRESS(type, actual_size);                        \
+            break;                                                                          \
+        }                                                                                   \
+        default:                                                                            \
+            break;                                                                          \
+        }                                                                                   \
+    } while (false)
+
+static inline void updatePartitionWriterMetrics(CompressionMethod method, size_t original_size, size_t actual_size, bool is_local)
 {
-    // statistic
-    GET_METRIC(tiflash_exchange_data_bytes, type_hash_original).Increment(packet_bytes);
-    // compression method is always NONE
-    if (is_local)
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_local).Increment(packet_bytes);
-    else
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_remote).Increment(packet_bytes);
+    UPDATE_EXCHANGE_MATRIC(hash, method, original_size, actual_size, is_local);
 }
 
-void updatePartitionWriterMetrics(CompressionMethod method, size_t original_size, size_t sz, bool is_local)
+template <bool is_broadcast, typename FuncIsLocalTunnel, typename FuncWriteToTunnel>
+static void broadcastOrPassThroughWriteImpl(
+    const size_t tunnel_cnt,
+    const size_t local_tunnel_cnt, // can be 0 for PassThrough writer
+    size_t ori_packet_bytes, // original data packet size
```

Comment:
It seems unnecessary to pass `ori_packet_bytes`.
Why not `auto local_tracked_packet_bytes = local_tracked_packet? local_tracked_packet->getPacket().ByteSizeLong() : 0;`?

Reply:
`ori_packet_bytes` is used for present uncompressed size of data packet. Especially when `local_tunnel_cnt` is 0 and `local_tracked_packet` is NULL, which may occur when using Passthrough.